lib/ide/src/view/text_editor.rs

```diff
@@ -2,36 +2,70 @@
 
 use crate::prelude::*;
 
-use super::temporary_panel::TemporaryPanel;
-use super::temporary_panel::TemporaryPadding;
+use crate::view::notification::NotificationService;
+use crate::view::temporary_panel::TemporaryPadding;
+use crate::view::temporary_panel::TemporaryPanel;
 
 use basegl::display::object::DisplayObjectOps;
 use basegl::display::shape::text::glyph::font::FontRegistry;
 use basegl::display::shape::text::text_field::TextField;
 use basegl::display::shape::text::text_field::TextFieldProperties;
 use basegl::display::world::*;
-
+use enso_frp::io::Key;
+use enso_frp::io::KeyboardActions;
+use enso_frp::io::KeyMask;
 use nalgebra::Vector2;
 use nalgebra::zero;
 
 
+
 // ==================
 // === TextEditor ===
 // ==================
 
+shared! { TextEditor
+
 /// TextEditor allows us to edit text files or Enso Modules. Extensible code highlighting is
 /// planned to be implemented for it.
-#[derive(Clone,Debug)]
-pub struct TextEditor {
-    text_field : TextField,
-    padding    : TemporaryPadding,
-    position   : Vector2<f32>,
-    size       : Vector2<f32>
+#[derive(Debug)]
+pub struct TextEditorData {
+    text_field           : TextField,
+    padding              : TemporaryPadding,
+    position             : Vector2<f32>,
+    size                 : Vector2<f32>,
+    controller           : controller::text::Handle,
+    notification_service : NotificationService
 }
 
+impl {
+    /// Saves text editor's content to file.
+    pub fn save(&self) {
+        let controller   = self.controller.clone();
+        let file_path    = controller.file_path();
+        let text         = self.text_field.get_content();
+        let store_fut    = controller.store_content(text);
+        let notification = self.notification_service.clone();
+        let duration     = 1.0;
+        let fade_out     = 1.0;
+        notification.info("Saving file", duration, fade_out);
```

Comment:
Client of notification service should never control the apperance. Apperance for "info" should be defined in the notification service.

One more thing - again important - I really, really hate notifications like "saving a file". This just makes the GUI ugly. I believe that notifications should be very very rare in the GUI - only in case of serious error. In normal workflow, I do not want our GUI to display **any notifications**. The GUI should be clean and do not disturb constantly with popups. Only warnings and errors duplicated from the logger.